### ISSUE TITLE
The saved best.pt actually corresponds to the previous epoch of the true best fitness.

### ISSUE THREAD

**Issue by blueOf** [ultralytics/ultralytics#23464]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

Actually, I set the save interval to 1 and record both best_fitness and fitness at the end of every epoch. After adding code to save the fitness and print the current epoch, I found a misalignment issue."

ultralytics/engine/trainer.py -754

<img width="822" height="98" alt="Image" src="https://github.com/user-attachments/assets/833d5daf-5f77-45ad-bf50-b6db1c5480da" />

result:
 Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
 405/650      11.2G      1.198     0.9202      1.085        191        640: 100% ━━━━━━━━━━━━ 94/94 4.2it/s 22.3s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 8/8 3.0it/s 2.7s
                   all       1000       2612       0.74      0.587      0.652      0.388
 🏆 --------------------epoch  404:

### Environment

Ultralytics 8.4.8 🚀 Python-3.10.19 torch-2.5.1 CUDA:0 (NVIDIA RTX A5000, 24123MiB)
Setup complete ✅ (256 CPUs, 1007.7 GB RAM, 2477.5/14193.6 GB disk)

OS                     Linux-5.15.0-139-generic-x86_64-with-glibc2.31
Environment            Linux
Python                 3.10.19
Install                pip
Path                   /home/nkq/anaconda3/envs/yolo11/lib/python3.10/site-packages/ultralytics
RAM                    1007.66 GB
Disk                   2477.5/14193.6 GB
CPU                    AMD EPYC 7763 64-Core Processor
CPU count              256
GPU                    NVIDIA RTX A5000, 24123MiB
GPU count              8
CUDA                   12.1

numpy                  ✅ 1.26.4>=1.23.0
matplotlib             ✅ 3.10.1>=3.3.0
opencv-python          ✅ 4.12.0.88>=4.6.0
pillow                 ✅ 11.1.0>=7.1.2
pyyaml                 ✅ 6.0.3>=5.3.1
requests               ✅ 2.32.5>=2.23.0
scipy                  ✅ 1.15.2>=1.4.1
torch                  ✅ 2.5.1<2.10,>=1.8.0
torch                  ✅ 2.5.1!=2.4.0,<2.10,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.20.1>=0.9.0
psutil                 ✅ 7.2.1>=5.8.0
polars                 ✅ 1.36.1>=0.20.0
ultralytics-thop       ✅ 2.0.18>=2.0.18

### Minimal Reproducible Example

just any datasets

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @blueOf, thank you for your interest in Ultralytics 🚀! This is an automated response to help us triage your report, and an Ultralytics engineer will also assist soon. In the meantime, we recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report (your report about `best.pt` appearing to lag one epoch vs. the printed best fitness), please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. “Any dataset” is a good start, but to reproduce reliably we still need a small, concrete MRE including the exact command/script and the smallest dataset/config that shows the misalignment.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version (please paste the new `yolo checks` / environment summary after upgrading) 🔄:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

To help the engineer quickly validate the suspected epoch/fitness misalignment, please add the following to your MRE details (copy/paste is fine) 🧩:
-Exact `yolo train ...` CLI command or Python snippet used (including `save_period=1`, `val`, `patience`, etc.)
-Whether this is single-GPU or multi-GPU (DDP) for the run you observed (you have 8 GPUs available)
-The full training log lines around the “new best” event and the corresponding checkpoint write (a small window of epochs is enough)
-Your exact edits around `ultralytics/engine/trainer.py` (line ~754) shown in the screenshot, or a minimal diff

**Comment by blueOf** [comment]
python valid.py runs/train/yolo11/weights/epoch336.pt
SVRDD/test 验证结果:
  mAP50:     0.6549
  mAP50-95:  0.3894
  Precision: 0.6869
  Recall:    0.6095
  F1-Score:  0.6459
  FPS:       165.41

python valid.py runs/train/yolo11/weights/best.pt
SVRDD/test 验证结果:
  mAP50:     0.6549
  mAP50-95:  0.3894
  Precision: 0.6869
  Recall:    0.6095
  F1-Score:  0.6459
  FPS:       164.91

python valid.py runs/train/yolo11/weights/epoch337.pt
SVRDD/test 验证结果:
  mAP50:     0.6551
  mAP50-95:  0.3897
  Precision: 0.6856
  Recall:    0.6093
  F1-Score:  0.6452
  FPS:       162.68


result.csv 
336,16928.8,1.27065,1.01636,1.12222,0.72738,0.58895,0.65119,0.39079,1.38012,1.17583,1.23272,0.00481085,0.00481085,0.00481085,0.39079,0.39112
337,16979.3,1.25975,1.01415,1.11959,0.72818,0.58737,0.65093,0.39126,1.38011,1.17556,1.23313,0.00478696,0.00478696,0.00478696,0.39126,0.39126
The last two columns are the current fitness (i.e., mAP@0.5:0.95) and best_fitness, respectively

**Comment by Y-T-G** [comment]
`self.epoch` counts from 0

Epoch 336 in CSV is `epoch337.pt`
